xc/xc7/techmap/cells_map.v

```
@@ -2496,33 +2566,33 @@ module OSERDESE2 (
   );
 
   parameter DATA_RATE_OQ = "DDR";
-  parameter DATA_RATE_TQ = "DDR";
+  parameter DATA_RATE_TQ = "BUF";
   parameter DATA_WIDTH = 4;
   parameter SERDES_MODE = "MASTER";
-  parameter TRISTATE_WIDTH = 4;
+  parameter TRISTATE_WIDTH = 1;
 
   if (DATA_RATE_OQ == "DDR" &&
       !(DATA_WIDTH == 2 || DATA_WIDTH == 4 ||
         DATA_WIDTH == 6 || DATA_WIDTH == 8)) begin
-    wire _TECHMAP_FAIL_ = 1'b1;
+    wire _TECHMAP_FAIL_;
```

Why did you change the _TECHMAP_FAIL_ here?  Has yosys's behavior around this feature changed?

---

Yosys behavior hasn't changed. The thing is when you declare the `_TECHMAP_FAIL_` wire and assign it `1` then the tehcmap fails silently and the un-mapped cell remains. But when you don't assign it anything then the Yosys throws an error. The former is a way of telling Yosys that a techmap is not applicable. I think that the latter is more what we want here (and probably everywhere).

---

I think you are depending on undocumented yosys behavior.  If you want to generate an error, I believe yosys has now implemented `$error`

---

This is not undocumented behavior, see manual for the `techmap` command:
```
    _TECHMAP_FAIL_
        When this wire is set to a non-zero constant value, techmap will not
        use this module and instead try the next module with a matching
        'techmap_celltype' attribute.

        When such a wire exists but does not have a constant value after all
        _TECHMAP_DO_* commands have been executed, an error is generated.
```

---

You didn't set a value at all.  Yosys generating an error when `_TECHMAP_FAIL_` has no value is an undocumented behavior.  The previous code set `_TECHMAP_FAIL_` to 1, which is non-zero, which what the documentation called for.  You are depending on undocumented behavior.

---

I disagree. What I'm doing is exactly what the documentation says in the second paragraph. See the example below. I've deliberately set parameters of a `OSERDES` to trigger the techmap failure. When there is `wire _TECHMAP_FAIL_ = 1` then the techmap fails silently and we end up with unmapped `OSERDESE2` in EBLIF:
```
   Number of cells:                  8
     BUFGCTRL_VPR                    2
     IBUF_VPR                        3
     OBUFT_VPR                       1
     OSERDESE2                       1
     T_INV                           1
```

But when I don't assign the wire with anything then Yosys fails and prints out the error cause which is exactly what I intended it to do:
```
ERROR: Techmap yielded config wire _TECHMAP_FAIL_ with non-const value \_TECHMAP_FAIL_.
```

Besides that I realzed that some of these error checks for OSERDES are incorrect. I'll file an issue about that.

---

I agree, resolving